### PR TITLE
[Bug](cast) change binary predicate finally cast to varchar

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
# Proposed changes
The phone is largeint here.
```sql
mysql> select * from test_demo where phone='195456789345678955654444443878';
+---------+--------------------------------+
| user_id | phone                          |
+---------+--------------------------------+
| 10003   | 195456789345678955654444443878 |
| 10004   | 195456789345678955654444443878 |
| 10005   | 195456789345678955654444443877 |
| 10006   | 195456789345678955654444443878 |
| 10009   | 195456789345678955654444443878 |
+---------+--------------------------------+
```
## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/analysis/BinaryPredicate.java [hunk 1]
```diff
@@ -391,7 +391,7 @@ private Type getCmpType() throws AnalysisException {
             }
         }
 
-        return Type.DOUBLE;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/BinaryPredicate.java:394]: Why no LARGEINT?

Comment (BiteTheDDDDt) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/BinaryPredicate.java]: This case will cast to LARGEINT now.

File: fe/fe-core/src/main/java/org/apache/doris/analysis/BinaryPredicate.java [hunk 2]
```diff
@@ -376,18 +376,15 @@ private Type getCmpType() throws AnalysisException {
         // When int column compares with string, Mysql will convert string to int.
         // So it is also compatible with Mysql.
 
-        if (t1 == PrimitiveType.BIGINT && (t2 == PrimitiveType.VARCHAR || t2 == PrimitiveType.STRING)) {
-            Expr rightChild = getChild(1);
-            Long parsedLong = Type.tryParseToLong(rightChild);
-            if (parsedLong != null) {
+        if (t1.isStringType() || t2.isStringType()) {
+            if (t1 == PrimitiveType.BIGINT && Type.canParseToBigInt(getChild(1))
+                    || t2 == PrimitiveType.BIGINT && Type.canParseToBigInt(getChild(0))) {
```
Comment (Gabriel39) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/BinaryPredicate.java:381]: if ((t1 == PrimitiveType.BIGINT && Type.canParseToBigInt(getChild(1)))
                    || (t2 == PrimitiveType.BIGINT && Type.canParseToBigInt(getChild(0))))

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]
LGTM

**Pull request merged**